content/300-guides/300-upgrade-guides/200-upgrading-versions/600-upgrading-to-prisma-5.mdx

```diff
@@ -0,0 +1,677 @@
+4. Review changes between your schema backup and the new schema generated by `db pull`. You can either use the new schema as is, or update it to include your preferred spacing, comments, etc.
+5. delete your existing migrations. We will be [performing a baseline](/getting-started/setup-prisma/add-to-existing-project/relational-databases/baseline-your-database-typescript-cockroachdb) in order to make your local setup agree with your existing CockroachDB instance.
+6. Perform the [baselining steps](/getting-started/setup-prisma/add-to-existing-project/relational-databases/baseline-your-database-typescript-cockroachdb). After these steps, you'll have migrated successfully from the `postgresql` provider to the `cockroachdb` provider!
+
+### Removal of <inlinecode>runtime/index.js</inlinecode> from generated client
+
+The `runtime/index.js` file has been removed from the Prisma Client.
+
+#### Using public APIs from <inlinecode>runtime/index.js</inlinecode>
+
+Importing from `@prisma/client/runtime` is no longer available in Prisma 5. If you were using public APIs available in this namespace before, you can instead import `Prisma` and access them. For example:
+
+```js
+import { Decimal } from '@prisma/client/runtime'
+const num = new Decimal(24.454545)
+```
+
+will need to be changed to
+
+```js
+import { Prisma } from '@prisma/client'
+const num = new Prisma.Decimal(24.454545)
+```
+
+#### Using private APIs for a specific runtime
+
+We highly discourage the use of internal private APIs as they can change without warning and are not guaranteed to be supported. If your usage requires a private API that was previous available [please reach out to us on GitHub.](https://github.com/prisma/prisma/issues/new?assignees=&labels=&projects=&template=feature_request.md&title=)
+
+## Other changes
+
+The following changes may cause an application to initially throw an error message after upgrading to Prisma 5. Fortunately, they are easy to solve, as the underlying functionality has been removed for a while or the change is a simple string replace.
+
+### Removal of deprecated Prisma CLI flags
+
+Several deprecated CLI flags have been removed. All following flags are from previous APIs and are no longer needed:
+
+- `--preview-feature` used in `db execute`, `db seed`, and `db diff`
+- `--experimental` and `--early-access-feature` used in `migrate`
+- `--force`/`-f` used in `db push`
+- `--experimental-reintrospection` and `--clean` used in `db pull`
+
+The outdated use of `db push --force` can be replaced with the newer implementation `db push --accept-data-loss`.
+
+All other flags are from previous APIs and are no longer needed.
+
+### Removal of the <inlinecode>beforeExit</inlinecode> hook from the library engine
+
+The `beforeExit` hook has been removed from the Prisma library engine. Instead of this hook we recommend using built-in Node.js exit events.
```

Comment:
Should we add a clarification about why it was necessary for the binary engine in the first place and that's no longer relevant for the library engine?

Reply:
Adding more explanation is always good.

Can you make a suggestion what we would write @aqrln ? Then @jharrell can take that into account.

Reply:
With binary engine this hook implements functionality that wouldn't be possible otherwise: it allows users to run code right before the query engine process is killed due to the Node.js process preparing to exit. This may be useful to run one last db query for whatever reason for example. With library engine there's no process to kill, so Prisma's `beforeExit` hook provided no benefit at all over using something like `process.on('beforeExit')`, and its implementation was pure tech debt and had risks of memory leaks (it actually led to memory leaks in jest tests before we started using a `WeakMap` there, not sure if it had actual rather than potential problems in recent versions).

Reply:
Thank you for the write up @aqrln ! I'll take this and work it into the guide.